File: elevator_backend/elevator_system/core/dispatcher.py
from .elevator import Elevator
from .routing_strategy import RoutingStrategy, ScanRoutingStrategy, SustainableRoutingStrategy
from .types import StrategyType
from .request import Request

class Dispatcher:
    def __init__(self, num_elevators: int, min_floor: int = 0, max_floor: int = 9, strategy_type: StrategyType = StrategyType.SCAN):
        self.min_floor = min_floor
        self.max_floor = max_floor
        self.strategy_type = strategy_type
        
        self.elevators = [Elevator(id=i+1, min_floor=min_floor, max_floor=max_floor) for i in range(num_elevators)]
        self.strategy = self._create_strategy(strategy_type)
        
        self.step_count = 0
        self.logs = []
        self.all_requests = []

    def _create_strategy(self, strategy_type: StrategyType) -> RoutingStrategy:
        if strategy_type == StrategyType.SCAN:
            return ScanRoutingStrategy()
        elif strategy_type == StrategyType.SUSTAINABLE:
            return SustainableRoutingStrategy()
        return ScanRoutingStrategy()

    def switch_strategy(self, strategy_type: StrategyType):
        self.strategy_type = strategy_type
        self.strategy = self._create_strategy(strategy_type)
        self.logs.append(f"🔄 Strategy switched to: {self.strategy.name}")

    def add_request(self, current_floor: int, target_floor: int) -> Request:
        req = Request(current_floor, target_floor)
        self.all_requests.append(req)

        idx = self.strategy.assign_request(req, self.elevators)
        elevator = self.elevators[idx]
        elevator.assign_request(req)

        self.logs.append(f"📥 {str(req)} → assigned to {elevator.name}")
        return req

    def step(self):
        self.step_count += 1
        step_logs = [f"━━━ Step {self.step_count} ━━━"]

        for elevator in self.elevators:
            self.strategy.compute_direction(elevator)
            move_logs = elevator.step()
            step_logs.extend(move_logs)

        self.logs.extend(step_logs)

    def is_complete(self) -> bool:
        return all(e.is_idle for e in self.elevators)

    def snapshot(self) -> dict:
        return {
            "stepCount": self.step_count,
            "strategyName": self.strategy.name,
            "strategyDescription": self.strategy.description,
            "elevators": [e.snapshot() for e in self.elevators],
            "isComplete": self.is_complete(),
            "totalRequests": len(self.all_requests),
            "servicedRequests": len([r for r in self.all_requests if r.serviced]),
            "pendingRequests": len([r for r in self.all_requests if not r.serviced and not r.picked_up]),
            "inTransit": len([r for r in self.all_requests if r.picked_up and not r.serviced]),
            "logs": self.logs.copy(),
            "allRequests": [r.to_dict() for r in self.all_requests]
        }

File: src/engine/Dispatcher.ts
import { Elevator, ElevatorSnapshot } from "./Elevator";
import { Request } from "./Request";
import { RoutingStrategy, ScanRoutingStrategy, SustainableRoutingStrategy } from "./RoutingStrategy";
import { StrategyType } from "./types";

/**
 * Central Dispatcher — Orchestrates the entire elevator system.
 *
 * Responsibilities:
 *  1. Maintains the pool of Elevator instances
 *  2. Receives incoming passenger Requests
 *  3. Delegates assignment to the active RoutingStrategy
 *  4. Advances the simulation by one discrete time step
 *  5. Produces a complete system snapshot for the UI layer
 */
export class Dispatcher {
  public elevators: Elevator[];
  public strategy: RoutingStrategy;
  public strategyType: StrategyType;
  public stepCount: number = 0;
  public logs: string[] = [];
  public allRequests: Request[] = [];

  private readonly maxFloor: number;
  private readonly minFloor: number;

  constructor(
    numElevators: number,
    minFloor: number = 0,
    maxFloor: number = 9,
    strategyType: StrategyType = StrategyType.SCAN
  ) {
    this.minFloor = minFloor;
    this.maxFloor = maxFloor;
    this.strategyType = strategyType;

    // Initialize elevators at ground floor
    this.elevators = [];
    for (let i = 0; i < numElevators; i++) {
      this.elevators.push(new Elevator(i + 1, 0, minFloor, maxFloor));
    }

    // Initialize with default strategy
    this.strategy = this.createStrategy(strategyType);
  }

  /** Factory method for strategy creation. */
  private createStrategy(type: StrategyType): RoutingStrategy {
    switch (type) {
      case StrategyType.SCAN:
        return new ScanRoutingStrategy();
      case StrategyType.SUSTAINABLE:
        return new SustainableRoutingStrategy();
    }
  }

  /** Hot-swap the routing strategy at runtime. */
  switchStrategy(type: StrategyType): void {
    this.strategyType = type;
    this.strategy = this.createStrategy(type);
    this.logs.push(`🔄 Strategy switched to: ${this.strategy.name}`);
  }

  /**
   * Submit a new passenger request.
   * The strategy decides which elevator receives it.
   */
  addRequest(currentFloor: number, targetFloor: number): Request {
    const request = new Request(currentFloor, targetFloor);
    this.allRequests.push(request);

    const elevatorIdx = this.strategy.assignRequest(
      request,
      this.elevators
    );
    const elevator = this.elevators[elevatorIdx];
    elevator.assignRequest(request);

    this.logs.push(
      `📥 ${request.toString()} → assigned to ${elevator.name}`
    );

    return request;
  }

  /**
   * Advance the simulation by one discrete time step.
   *
   * Each step:
   *  1. Compute direction for each elevator (via strategy)
   *  2. Execute one move per elevator
   *  3. Collect logs
   */
  step(): StepResult {
    this.stepCount++;
    const stepLogs: string[] = [`━━━ Step ${this.stepCount} ━━━`];

    for (const elevator of this.elevators) {
      // Strategy decides direction before each move
      this.strategy.computeDirection(elevator);

      // Elevator executes one step
      const moveLogs = elevator.step();
      stepLogs.push(...moveLogs);
    }

    this.logs.push(...stepLogs);

    return {
      step: this.stepCount,
      logs: stepLogs,
      snapshots: this.elevators.map((e) => e.snapshot()),
      isComplete: this.isComplete(),
    };
  }

  /** Check if all requests have been serviced. */
  isComplete(): boolean {
    return this.elevators.every((e) => e.isIdle);
  }

  /** Reset the entire system to initial state. */
  reset(): void {
    this.stepCount = 0;
    this.logs = [];
    this.allRequests = [];
    const numElevators = this.elevators.length;
    this.elevators = [];
    for (let i = 0; i < numElevators; i++) {
      this.elevators.push(
        new Elevator(i + 1, 0, this.minFloor, this.maxFloor)
      );
    }
  }

  /** Full system snapshot for rendering. */
  snapshot(): SystemSnapshot {
    return {
      stepCount: this.stepCount,
      strategyName: this.strategy.name,
      strategyDescription: this.strategy.description,
      elevators: this.elevators.map((e) => e.snapshot()),
      isComplete: this.isComplete(),
      totalRequests: this.allRequests.length,
      servicedRequests: this.allRequests.filter((r) => r.serviced).length,
      pendingRequests: this.allRequests.filter(
        (r) => !r.serviced && !r.pickedUp
      ).length,
      inTransit: this.allRequests.filter(
        (r) => r.pickedUp && !r.serviced
      ).length,
    };
  }
}

export interface StepResult {
  step: number;
  logs: string[];
  snapshots: ElevatorSnapshot[];
  isComplete: boolean;
}

export interface SystemSnapshot {
  stepCount: number;
  strategyName: string;
  strategyDescription: string;
  elevators: ElevatorSnapshot[];
  isComplete: boolean;
  totalRequests: number;
  servicedRequests: number;
  pendingRequests: number;
  inTransit: number;
}
